fastfield_codecs/src/compact_space.rs

```diff
@@ -0,0 +1,683 @@
+use std::{
+    cmp::Ordering,
+    collections::BTreeSet,
+    io::{self, Write},
+    net::{IpAddr, Ipv6Addr},
+    ops::RangeInclusive,
+};
+
+use common::{BinarySerializable, CountingWriter, VIntU128};
+use ownedbytes::OwnedBytes;
+use tantivy_bitpacker::{self, BitPacker, BitUnpacker};
+
+use crate::column::{ColumnV2, ColumnV2Ext};
+use crate::compact_space::build_compact_space::get_compact_space;
+
+mod blank_range;
+mod build_compact_space;
+
+pub fn ip_to_u128(ip_addr: IpAddr) -> u128 {
+    let ip_addr_v6: Ipv6Addr = match ip_addr {
+        IpAddr::V4(v4) => v4.to_ipv6_mapped(),
+        IpAddr::V6(v6) => v6,
+    };
+    u128::from_be_bytes(ip_addr_v6.octets())
+}
+
+/// The cost per blank is quite hard actually, since blanks are delta encoded, the actual cost of
+/// blanks depends on the number of blanks.
+///
+/// The number is taken by looking at a real dataset. It is optimized for larger datasets.
+const COST_PER_BLANK_IN_BITS: usize = 36;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct CompactSpace {
+    ranges_and_compact_start: Vec<(RangeInclusive<u128>, u64)>,
+    pub null_value: u128,
+}
+
+impl BinarySerializable for CompactSpace {
+    fn serialize<W: io::Write>(&self, writer: &mut W) -> io::Result<()> {
+        VIntU128(self.null_value).serialize(writer)?;
+        VIntU128(self.ranges_and_compact_start.len() as u128).serialize(writer)?;
+
+        let mut prev_value = 0;
+        for (value_range, _compact) in &self.ranges_and_compact_start {
+            let blank_delta_start = value_range.start() - prev_value;
+            VIntU128(blank_delta_start).serialize(writer)?;
+            prev_value = *value_range.start();
+
+            let blank_delta_end = value_range.end() - prev_value;
+            VIntU128(blank_delta_end).serialize(writer)?;
+            prev_value = *value_range.end();
+        }
+
+        Ok(())
+    }
+
+    fn deserialize<R: io::Read>(reader: &mut R) -> io::Result<Self> {
+        let null_value = VIntU128::deserialize(reader)?.0;
+        let num_values = VIntU128::deserialize(reader)?.0;
+        let mut ranges_and_compact_start: Vec<(RangeInclusive<u128>, u64)> = vec![];
+        let mut value = 0u128;
+        let mut compact = 0u64;
+        for _ in 0..num_values {
+            let blank_delta_start = VIntU128::deserialize(reader)?.0;
+            value += blank_delta_start;
+            let blank_start = value;
+
+            let blank_delta_end = VIntU128::deserialize(reader)?.0;
+            value += blank_delta_end;
+            let blank_end = value;
+
+            let compact_delta = blank_end - blank_start + 1;
+
+            ranges_and_compact_start.push((blank_start..=blank_end, compact));
+            compact += compact_delta as u64;
+        }
+
+        Ok(Self {
+            null_value,
+            ranges_and_compact_start,
+        })
+    }
+}
+
+impl CompactSpace {
+    fn amplitude_compact_space(&self) -> u128 {
```

Comment:
I am not sure how you defined amplitude here so I cannot readproof this.

Can you add a comment here.

e.g.
Amplitude is a number such that the compact space consists including the sentinel value used to identify null values.
 
The compact space is `0..=amplitude` .
```

Reply:
Added a comment, which is pretty much what you proposed